### PR TITLE
mon: use v2 addrs for monmap, bootstrap, etc

### PR DESCRIPTION
**Pull request by liewegas**
This includes

- addrs now print with either a v1: or v2: prefix for clarity
- a ton of little fixes to ProtocolV2
- changes to monmap to use addrvec
- daemons bind to v1 and v2 addresses now by default
- a zillion changes to the bootstrap process to handle that fact.
- building a monmap for discovering a mon vs building a map to build a cluster now have different behaviors (the former tries v1 and v2 protocols more liberally, the latter has to make assumptions about what an ip:port really means)
- some mds fixes to handle addrs
- some random stuff in teuthology.git misc.py is moved into ceph.py, where i can change it
- teuthology ceph task gets some new options to control whether we do v2 addresses or not
- upgrade tests don't start with v2 addresses, obviously
- teuthology ceph.conf now uses mon host = ... in [global] instead of the old mon addr options in [mon.a] etc.

TODO
- [x] make legacy json fields leave off the v1 prefix (i.e., avoid breaking things like cinder)
- [x] MonClient: pivot to v2 address if available for current mon connection
- [x] osd,mds,mgr: only bind to v2 if monmap has NAUTILUS feature

MAYBE
- [ ] osd: pass addresses in boot message
- [ ] mds: pass addresses in beacon message

### CODE REVIEW

File: src/msg/async/AsyncConnection.cc [hunk 1]
```diff
@@ -462,16 +465,19 @@ void AsyncConnection::_connect()
...
+			     << " peer_addr " << peer_addr << dendl;
   ceph_assert(socket.fd() >= 0);
 
   std::lock_guard<std::mutex> l(lock);
   cs = std::move(socket);
-  socket_addr = addr;
-  target_addr = addr; // until we know better
+  socket_addr = listen_addr;
+  target_addr = peer_addr; // until we know better
   state = STATE_ACCEPTING;
```
Comment (runsisi) [inline src/msg/async/AsyncConnection.cc]: @liewegas i think we should change line 494 from `m->get_connection().get()` to `this` ?

File: src/mon/MonMap.cc [hunk 2]
```diff
@@ -422,11 +425,11 @@ void MonMap::set_initial_members(CephContext *cct,
...
   // add missing initial members
-  for (list<string>::iterator p = initial_members.begin(); p != initial_members.end(); ++p) {
-    if (!contains(*p)) {
-      if (*p == my_name) {
-	lgeneric_dout(cct, 1) << " adding self " << *p << " " << my_addr << dendl;
-	add(*p, my_addr);
+  for (auto& p : initial_members) {
+    if (!contains(p)) {
+      if (p == my_name) {
+	lgeneric_dout(cct, 1) << " adding self " << p << " " << my_addr << dendl;
```
Comment (neha-ojha) [inline src/mon/MonMap.cc]: should be `my_addrs`

File: src/mon/MonMap.cc [hunk 3]
```diff
@@ -339,6 +339,39 @@ void MonMap::dump(Formatter *f) const
   f->close_section();
 }
 
+// an ambiguous mon addr may be legacy or may be msgr2--we aren' sure.
```
Comment (neha-ojha) [inline src/mon/MonMap.cc]: nit: aren't

File: src/mon/MonMap.cc (hunk 2 above)
Comment (neha-ojha) [reply src/mon/MonMap.cc]: nvm, has been fixed in a later commit.

File: src/msg/msg_types.cc [hunk 4]
```diff
@@ -172,10 +177,8 @@ ostream& operator<<(ostream& out, const entity_addr_t &addr)
   if (addr.type == entity_addr_t::TYPE_NONE) {
     return out << "-";
   }
-  if (addr.type != entity_addr_t::TYPE_DEFAULT) {
```
[the remark this answers is not in the document]
Comment (liewegas) [reply src/msg/msg_types.cc]: It does.  in most cases I left the old field in place with the legacy address, and added a new 'addrs' (or whatever it was, but plural) and dump the addrvec.  I can make it leave off the prefix, but eventually we'll want to adjust the openstack code to understand the v2 and v1 addresses.

File: src/msg/msg_types.cc [hunk 5]
```diff
@@ -62,25 +62,30 @@ void entity_inst_t::generate_test_instances(list<entity_inst_t*>& o)
...
+  if (strncmp("v1:", s, 3) == 0) {
+    start += 3;
     newtype = TYPE_LEGACY;
-  } else if (strncmp("msgr2:", s, 6) == 0) {
-    start += 6;
+  } else if (strncmp("v2:", s, 3) == 0) {
+    start += 3;
     newtype = TYPE_MSGR2;
+  } else if (strncmp("any:", s, 4) == 0) {
+    start += 3;
```
Comment (jdurgin) [inline src/msg/msg_types.cc]: start += 4

File: src/mon/OSDMonitor.cc [hunk 6]
```diff
@@ -10178,6 +10178,12 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
       goto reply;
     }
     if (rel == CEPH_RELEASE_MIMIC) {
+      if (!mon->monmap->get_required_features().contains_all(
+	    ceph::features::mon::FEATURE_MIMIC)) {
```
Comment (jdurgin) [inline src/mon/OSDMonitor.cc]: why is this needed? hope it's not too important since existing mimic won't have it

**Comment by jdurgin** [comment]
should we suggest avoiding 'mon addr' in the docs too?

Several places mention it:
man/8/ceph-mon.rst:     host1,host2,...``), or ``mon addr`` lines in *ceph.conf*.  If this
man/8/monmaptool.rst:      #. ``[mon.foo]`` sections containing ``mon addr`` settings in the config
rados/configuration/common.rst: mon addr = 150.140.130.120:6789
rados/configuration/mon-config-ref.rst:         mon addr = 10.0.0.10:6789,10.0.0.11:6789,10.0.0.12:6789
rados/configuration/mon-config-ref.rst:         mon addr = 10.0.0.10:6789
rados/configuration/network-config-ref.rst:             mon addr = {ip-address}:6789
rados/operations/add-or-rm-mons.rst:   ``--public-addr {ip:port}`` or by setting ``mon addr`` in the
start/ceph.conf:        mon addr = {ip-address}:6789

**Comment by jdurgin** [comment]
docs about firewall config/default ports should mention 3300 now, e.g. ```start/quick-start-preflight.rst
rados/configuration/network-config-ref.rst
rados/configuration/common.rst
```
Perhaps a section explaining v1 and v2 and multibind options in the network config ref would help too.

**Comment by liewegas** [comment]
i don't think we need to worry about the skeleton config stuff if we leave it in place in teuthology.git for 2 releases.

ugh, the docs are so old and so wrong.. they still say you have to define 'host' options, for example.  grr.

**Comment by liewegas** [comment]
@tchaikov i haven't gotten my local builds to do seastar yet, so i'm relying on make check to fix the errors.

**Comment by tchaikov** [comment]
@liewegas i will try to rebase on your PR to build and exercise the crimson tests also.

File: doc/rados/configuration/network-config-ref.rst [hunk 7]
```diff
@@ -15,10 +15,13 @@ cluster network, Ceph assumes a single "public" network. Ceph functions just
...
 
-We recommend running a Ceph Storage Cluster with two networks: a public
-(front-side) network and a cluster (back-side) network. To support two networks,
-each :term:`Ceph Node` will need to have more than one NIC. See `Hardware
-Recommendations -  Networks`_ for additional details.
+It is possible to run a Ceph Storage Cluster with two networks: a public
+(front-side) network and a cluster (back-side) network.  However, this approach
+complicates network configuration (both hardware and software) and does not usually have a significant impact on overall performance.  For this reason, we generally recommend that dual-NIC systems either be configured with two IPs on the same network, or bonded.
+
+If, despite the complexity, on still wishes to use two networks, each
```
Comment (jdurgin) [inline doc/rados/configuration/network-config-ref.rst]: s/ on / one /

File: doc/man/8/monmaptool.rst [hunk 8]
```diff
@@ -53,7 +53,7 @@ Options
 
       #. ``--monmap filename`` to specify a monmap to load
       #. ``--mon-host 'host1,ip2'`` to specify a list of hosts or ip addresses
-      #. ``[mon.foo]`` sections containing ``mon addr`` settings in the config
+      #. ``[mon.foo]`` sections containing ``mon addr`` settings in the config. Note that this method is not recommended and support will be remove in a future release.
```
Comment (jdurgin) [inline doc/man/8/monmaptool.rst]: s/remove/removed/

**Comment by jdurgin** [comment]
encoding entity_addr_t for old clients makes sense. It makes me wonder if the changes to operator<< for entity_addr_t may cause any issues for rgw/rbd - @dillaman @yehudasa @cbodley anything that might be impacted by changing that during an upgrade?

**Comment by cbodley** [comment]
@jdurgin i wouldn't expect any problems with rgw, as the messenger internals are hidden from us under the librados client

**Comment by liewegas** [comment]
retest this please

**Comment by liewegas** [comment]
retest this please

**Comment by dillaman** [comment]
@liewegas Unfortunately it came from play testing in downstream configuration against an upstream cluster for iSCSI.

**Comment by liewegas** [comment]
ok i'll try to reproduce locally!  thanks

**Comment by liewegas** [comment]
@dillaman https://github.com/ceph/ceph/pull/25948
